src/tools/tidy/src/ext_tool_checks.rs

```diff
@@ -39,6 +40,85 @@ const BLACK_CONFIG_PATH: &[&str] = &["src", "tools", "tidy", "config", "black.to
 const RUFF_CACH_PATH: &[&str] = &["cache", "ruff_cache"];
 const PIP_REQ_PATH: &[&str] = &["src", "tools", "tidy", "config", "requirements.txt"];
 
+#[derive(Debug, PartialEq)]
+pub enum DiffLine {
+    Context(String),
+    Expected(String),
+    Resulting(String),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Mismatch {
+    pub line_number: u32,
+    pub lines: Vec<DiffLine>,
+}
+
+impl Mismatch {
+    fn new(line_number: u32) -> Mismatch {
+        Mismatch { line_number, lines: Vec::new() }
+    }
+}
+
+// Produces a diff between the expected output and actual output.
+pub fn make_diff(expected: &str, actual: &str, context_size: usize) -> Vec<Mismatch> {
```

Comment:
I copied it from https://github.com/rust-lang/rust/blob/d929a42a664c026167800801b26d734db925314f/src/tools/rustfmt/src/rustfmt_diff.rs#L180. Besides submit this part of the code to upstream, I don't have any other good ideas.

Reply:
Could you copy the files to a temp directory, format them there, then just pass it to `diff -up`?

Reply:
I can just pass the expect content to stdin, for example: `clang-format compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp | diff -u compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp -`. But, this still requires the diff command line. I’m concerned about whether it can run properly on some Linux distributions (such as NixOS or environments with strict PATH restrictions) or on Windows.

Reply:
Ah yeah good point, probably better to go with Jakub’s solution